File: android/rctmgl/src/main/java-legacy/com/mapbox/rctmgl/components/camera/RCTMGLCamera.java
package com.mapbox.rctmgl.components.camera;

import android.content.Context;
import android.location.Location;

import com.mapbox.mapboxsdk.camera.CameraPosition;
import com.mapbox.mapboxsdk.camera.CameraUpdate;
import com.mapbox.mapboxsdk.camera.CameraUpdateFactory;
import com.mapbox.mapboxsdk.geometry.LatLng;
import com.mapbox.mapboxsdk.geometry.LatLngBounds;
import com.mapbox.mapboxsdk.geometry.VisibleRegion;
import com.mapbox.mapboxsdk.location.OnCameraTrackingChangedListener;
import com.mapbox.mapboxsdk.location.modes.CameraMode;
import com.mapbox.mapboxsdk.location.modes.RenderMode;
import com.mapbox.mapboxsdk.maps.MapboxMap;
import com.mapbox.mapboxsdk.maps.Style;
import com.mapbox.mapboxsdk.location.LocationComponent;
import com.mapbox.mapboxsdk.location.LocationComponentOptions;
import com.mapbox.mapboxsdk.location.LocationComponentActivationOptions;
// import com.mapbox.mapboxsdk.plugins.locationlayer.LocationLayerPlugin;
import com.mapbox.rctmgl.components.AbstractMapFeature;
import com.mapbox.rctmgl.components.location.LocationComponentManager;
import com.mapbox.rctmgl.components.mapview.RCTMGLMapView;
import com.mapbox.rctmgl.events.IEvent;
import com.mapbox.rctmgl.events.MapUserTrackingModeEvent;
import com.mapbox.rctmgl.events.MapChangeEvent;
import com.mapbox.rctmgl.location.LocationManager;
import com.mapbox.rctmgl.location.UserLocation;
import com.mapbox.rctmgl.location.UserLocationVerticalAlignment;
import com.mapbox.rctmgl.location.UserTrackingMode;
import com.mapbox.rctmgl.location.UserTrackingState;
import com.mapbox.rctmgl.utils.GeoJSONUtils;

import com.mapbox.rctmgl.R;

import com.mapbox.rctmgl.events.constants.EventTypes;

import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.WritableNativeMap;

import com.mapbox.geojson.Point;

import com.mapbox.rctmgl.impl.PermissionsManagerImpl;

import androidx.annotation.NonNull;

public class RCTMGLCamera extends AbstractMapFeature {
    private RCTMGLCameraManager mManager;
    private RCTMGLMapView mMapView;

    private boolean hasSentFirstRegion = false;

    private CameraStop mDefaultStop;
    private CameraStop mCameraStop;
    private CameraUpdateQueue mCameraUpdateQueue;

    // private LocationComponent mLocationComponent;
    private LocationComponentManager mLocationComponentManager;

    private int mUserTrackingMode;
    private int mUserTrackingState = UserTrackingState.POSSIBLE;
    private int mUserLocationVerticalAlignment = UserLocationVerticalAlignment.CENTER;

    public static final int USER_LOCATION_CAMERA_MOVE_DURATION = 1000;

    private LocationManager mLocationManager;
    private UserLocation mUserLocation;

    private Point mCenterCoordinate;

    private boolean mAnimated;
    private double mHeading;
    private double mPitch;
    private double mZoomLevel = -1;

    private double mMinZoomLevel = -1;
    private double mMaxZoomLevel = -1;

    private LatLngBounds mMaxBounds;

    private boolean mFollowUserLocation;
    private String mFollowUserMode;

    private Context mContext;


    private LocationManager.OnUserLocationChange mLocationChangeListener = new LocationManager.OnUserLocationChange() {
        @Override
        public void onLocationChange(Location nextLocation) {
        if (getMapboxMap() == null || mLocationComponentManager == null || !mLocationComponentManager.hasLocationComponent() || (!mFollowUserLocation)) {
            return;
        }

        mUserLocation.setCurrentLocation(nextLocation);
        sendUserLocationUpdateEvent(nextLocation);
        }
    };

    private MapboxMap.CancelableCallback mCameraCallback = new MapboxMap.CancelableCallback() {
        @Override
        public void onCancel() {
            if (!hasSentFirstRegion) {
                mMapView.sendRegionChangeEvent(false);
                hasSentFirstRegion = true;
            }
        }

        @Override
        public void onFinish() {
            if (!hasSentFirstRegion) {
                mMapView.sendRegionChangeEvent(false);
                hasSentFirstRegion = true;
            }
        }
    };

    public RCTMGLCamera(Context context, RCTMGLCameraManager manager) {
        super(context);
        mContext = context;
        mManager = manager;
        mCameraUpdateQueue = new CameraUpdateQueue();

        mUserLocation = new UserLocation();
        mLocationManager = LocationManager.getInstance(context);
    }

    @Override
    public void addToMap(RCTMGLMapView mapView) {
        mMapView = mapView;

        setInitialCamera();
        updateMaxMinZoomLevel();
        updateMaxBounds();

        if (mCameraStop != null) {
            updateCamera();
        }

        if (mFollowUserLocation) {
            enableLocation();
        }
    }

    @Override
    public void removeFromMap(RCTMGLMapView mapView) {

    }

    public void setStop(CameraStop stop) {
        mCameraStop = stop;
        mCameraStop.setCallback(mCameraCallback);

        if (mMapView != null) {
            updateCamera();
        }
    }

    public void setDefaultStop(CameraStop stop) {
        mDefaultStop = stop;
    }

    public void setFollowPitch(double pitch) {
        mPitch = pitch;
        updateCameraPositionIfNeeded(true);
    }

    public void setMaxBounds(LatLngBounds bounds) {
        mMaxBounds = bounds;
        updateMaxBounds();
    }

    private void updateMaxBounds() {
        MapboxMap map = getMapboxMap();
        if (map != null && mMaxBounds != null) {
            map.setLatLngBoundsForCameraTarget(mMaxBounds);
        }
    }

    private void updateMaxMinZoomLevel() {
        MapboxMap map = getMapboxMap();
        if (map != null) {
            if (mMinZoomLevel >= 0.0) {
                map.setMinZoomPreference(mMinZoomLevel);
            }
            if (mMaxZoomLevel >= 0.0) {
                map.setMaxZoomPreference(mMaxZoomLevel);
            }
        }
    }

    private void setInitialCamera() {
        if (mDefaultStop != null) {
            mDefaultStop.setDuration(0);
            mDefaultStop.setMode(com.mapbox.rctmgl.components.camera.constants.CameraMode.NONE);
            CameraUpdateItem item = mDefaultStop.toCameraUpdate(mMapView);
            item.run();
        }
    }

    private void updateCamera() {
        mCameraUpdateQueue.offer(mCameraStop);
        mCameraUpdateQueue.execute(mMapView);
    }

    private void updateUserTrackingMode(int userTrackingMode) {
        mUserLocation.setTrackingMode(userTrackingMode);
        IEvent event = new MapUserTrackingModeEvent(this, userTrackingMode);
        mManager.handleEvent(event);
    }

    private void updateUserLocation(boolean isAnimated) {
        if ((!mFollowUserLocation) || mUserLocation.getTrackingMode() == UserTrackingMode.NONE) {
            return;
        }

        if (mUserTrackingState == UserTrackingState.POSSIBLE) {
            updateUserLocationSignificantly(isAnimated);
        } else if (mUserTrackingState == UserTrackingState.CHANGED) {
            updateUserLocationIncrementally(isAnimated);
        }
    }

    private CameraPosition getUserLocationUpdateCameraPosition(double zoomLevel) {
        LatLng center = mUserLocation.getCoordinate();

        if (mUserLocationVerticalAlignment != UserLocationVerticalAlignment.CENTER) {
            VisibleRegion region = mMapView.getVisibleRegion(center, zoomLevel);

            switch (mUserLocationVerticalAlignment) {
                case UserLocationVerticalAlignment.TOP:
                    center = new LatLng(region.nearRight.getLatitude(), center.getLongitude());
                    break;
                case UserLocationVerticalAlignment.BOTTOM:
                    center = new LatLng(region.farLeft.getLatitude(), center.getLongitude());
                    break;
            }
        }

        return new CameraPosition.Builder()
                .target(center)
                .bearing(getDirectionForUserLocationUpdate())
                .tilt(mPitch)
                .zoom(zoomLevel)
                .build();
    }

    private double getDirectionForUserLocationUpdate() {
        // NOTE: The direction of this is used for map rotation only, not location layer rotation
        CameraPosition currentCamera = mMapView.getCameraPosition();
        double direction = currentCamera.bearing;

        int userTrackingMode = mUserLocation.getTrackingMode();
        if (userTrackingMode == UserTrackingMode.FollowWithHeading || userTrackingMode == UserTrackingMode.FollowWithCourse) {
            direction = mUserLocation.getBearing();
        } else if (mHeading != 0.0) {
            direction = mHeading;
        }

        return direction;
    }

    private void sendUserLocationUpdateEvent(Location location) {
        if(location == null){
            return;
        }
        IEvent event = new MapChangeEvent(this, EventTypes.USER_LOCATION_UPDATED, makeLocationChangePayload(location));
        mManager.handleEvent(event);
    }

    private boolean hasSetCenterCoordinate() {
        CameraPosition cameraPosition = mMapView.getCameraPosition();
        LatLng center = cameraPosition.target;
        return center.getLatitude() != 0.0 && center.getLongitude() != 0.0;
    }

    static final double minimumZoomLevelForUserTracking = 10.5;
    static final double defaultZoomLevelForUserTracking = 14.0;

    private void updateUserLocationSignificantly(boolean isAnimated) {
        mUserTrackingState = UserTrackingState.BEGAN;

        double zoom = mZoomLevel;
        if (zoom < 0) {
            double camerZoom = mMapView.getMapboxMap().getCameraPosition().zoom;
            if (camerZoom < minimumZoomLevelForUserTracking) {
                zoom = defaultZoomLevelForUserTracking;
            } else {
                zoom = camerZoom;
            }
        }
        CameraUpdate cameraUpdate = CameraUpdateFactory.newCameraPosition(getUserLocationUpdateCameraPosition(zoom));
        MapboxMap.CancelableCallback cameraCallback = new MapboxMap.CancelableCallback() {
            @Override
            public void onCancel() {
                mUserTrackingState = UserTrackingState.CHANGED;
            }

            @Override
            public void onFinish() {
                mUserTrackingState = UserTrackingState.CHANGED;
            }
        };

        if (isAnimated && hasSetCenterCoordinate()) {
            mMapView.animateCamera(cameraUpdate, cameraCallback);
        } else {
            mMapView.moveCamera(cameraUpdate, cameraCallback);
        }
    }

    private void updateUserLocationIncrementally(boolean isAnimated) {
        mUserTrackingState = UserTrackingState.BEGAN;

        CameraPosition cameraPosition = mMapView.getCameraPosition();
        CameraUpdate cameraUpdate = CameraUpdateFactory.newCameraPosition(getUserLocationUpdateCameraPosition(cameraPosition.zoom));

        MapboxMap.CancelableCallback callback = new MapboxMap.CancelableCallback() {
            @Override
            public void onCancel() {
                mUserTrackingState = UserTrackingState.CHANGED;
            }

            @Override
            public void onFinish() {
                mUserTrackingState = UserTrackingState.CHANGED;
            }
        };

        if (isAnimated) {
            mMapView.easeCamera(cameraUpdate, USER_LOCATION_CAMERA_MOVE_DURATION, true, callback);
        } else {
            mMapView.moveCamera(cameraUpdate, callback);
        }
    }

    private void enableLocation() {
        if (!PermissionsManagerImpl.areLocationPermissionsGranted(mContext)) {
            return;
        }

        if (!mLocationManager.isActive()) {
            mLocationManager.enable();
        }

        mMapView.getMapboxMap().getStyle(new Style.OnStyleLoaded() {
            @Override
            public void onStyleLoaded(@NonNull Style style) {
                enableLocationComponent(style);
            }
        });
    }

    private void enableLocationComponent(@NonNull Style style) {
        updateUserLocation(false);
        updateLocationLayer(style);

        Location lastKnownLocation = mLocationManager.getLastKnownLocation();
        mLocationManager.addLocationListener(mLocationChangeListener);

        if (lastKnownLocation != null) {
            mLocationChangeListener.onLocationChange(lastKnownLocation);

            postDelayed(new Runnable() {
                @Override
                public void run() {
                    mMapView.sendRegionDidChangeEvent();
                }
            }, 200);
        }

    }

    private void updateLocationLayer(@NonNull Style style) {
        if (mLocationComponentManager == null) {
            mLocationComponentManager = mMapView.getLocationComponentManager();
        }

        mLocationComponentManager.update(style);

        if (mFollowUserLocation) {
            mLocationComponentManager.setCameraMode(UserTrackingMode.getCameraMode(mUserTrackingMode));
        }
        mLocationComponentManager.setFollowUserLocation(mFollowUserLocation);

        if (mFollowUserLocation) {
            mLocationComponentManager.setCameraMode(UserTrackingMode.getCameraMode(mUserTrackingMode));
            mLocationComponentManager.addOnCameraTrackingChangedListener(new OnCameraTrackingChangedListener() {
                    @Override public void onCameraTrackingChanged(int currentMode) {
                        int userTrackingMode = UserTrackingMode.NONE;
                        switch (currentMode) {
                            case CameraMode.NONE:
                                userTrackingMode = UserTrackingMode.NONE;
                                break;
                            case CameraMode.TRACKING:
                                userTrackingMode = UserTrackingMode.FOLLOW;
                                break;
                            case CameraMode.TRACKING_COMPASS:
                                userTrackingMode = UserTrackingMode.FollowWithHeading;
                                break;
                            case CameraMode.TRACKING_GPS:
                                userTrackingMode = UserTrackingMode.FollowWithCourse;
                                break;
                            default:
                                userTrackingMode = UserTrackingMode.NONE;
                        }
                        updateUserTrackingMode(userTrackingMode);
                    }
                    @Override public void onCameraTrackingDismissed() {
                    }
            });
        } else {
            mLocationComponentManager.setCameraMode(CameraMode.NONE);
        }
    }

    public void setMinZoomLevel(double zoomLevel) {
        mMinZoomLevel = zoomLevel;
        updateMaxMinZoomLevel();
    }

    public void setMaxZoomLevel(double zoomLevel) {
        mMaxZoomLevel = zoomLevel;
        updateMaxMinZoomLevel();
    }

    public void setZoomLevel(double zoomLevel) {
        mZoomLevel = zoomLevel;
        updateCameraPositionIfNeeded(false);
    }

    private CameraPosition buildCamera(CameraPosition previousPosition, boolean shouldUpdateTarget) {
        CameraPosition.Builder builder = new CameraPosition.Builder(previousPosition)
                .bearing(mHeading)
                .tilt(mPitch)
                .zoom(mZoomLevel);

        if (shouldUpdateTarget) {
            builder.target(GeoJSONUtils.toLatLng(mCenterCoordinate));
        }

        return builder.build();
    }

    private void updateCameraPositionIfNeeded(boolean shouldUpdateTarget) {
        if (mMapView != null) {
            CameraPosition prevPosition = mMapView.getCameraPosition();
            CameraUpdate cameraUpdate = CameraUpdateFactory.newCameraPosition(buildCamera(prevPosition, shouldUpdateTarget));

            if (mAnimated) {
                mMapView.easeCamera(cameraUpdate);
            } else {
                mMapView.moveCamera(cameraUpdate);
            }
        }
    }

    public void setUserTrackingMode(int userTrackingMode) {
        int oldTrackingMode = mUserTrackingMode;
        mUserTrackingMode = userTrackingMode;
        updateUserTrackingMode(userTrackingMode);

        switch (mUserTrackingMode) {
            case UserTrackingMode.NONE:
                mUserTrackingState = UserTrackingState.POSSIBLE;
                break;
            case UserTrackingMode.FOLLOW:
            case UserTrackingMode.FollowWithCourse:
            case UserTrackingMode.FollowWithHeading:
                if (oldTrackingMode == UserTrackingMode.NONE) {
                    mUserTrackingState = UserTrackingState.POSSIBLE;
                }
                break;

        }

        if (getMapboxMap() != null) {
            updateLocationLayer(getMapboxMap().getStyle());
        }
    }


    public void setFollowUserLocation(boolean value) {
        mFollowUserLocation = value;
        updatedFollowUserMode();
    }

    public void setFollowUserMode(String mode) {
        mFollowUserMode = mode;
        updatedFollowUserMode();
    }

    private void updatedFollowUserMode() {
        if (mFollowUserLocation) {
            setUserTrackingMode(UserTrackingMode.fromString(mFollowUserMode));
        } else {
            setUserTrackingMode(UserTrackingMode.NONE);
        }
    }

    MapboxMap getMapboxMap() {
        if (mMapView == null) {
            return null;
        }
        return mMapView.getMapboxMap();
    }

    /**
     * Create a payload of the location data per the web api geolocation spec
     * https://dev.w3.org/geo/api/spec-source.html#position
     * @return
     */
    private WritableMap makeLocationChangePayload(Location location) {
        WritableMap positionProperties = new WritableNativeMap();
        WritableMap coords = new WritableNativeMap();

        coords.putDouble("longitude", location.getLongitude());
        coords.putDouble("latitude", location.getLatitude());
        coords.putDouble("altitude", location.getAltitude());
        coords.putDouble("accuracy", location.getAccuracy());
        // A better solution will be to pull the heading from the compass engine, 
        // unfortunately the api is not publicly available in the mapbox sdk
        coords.putDouble("heading", location.getBearing());
        coords.putDouble("course", location.getBearing());
        coords.putDouble("speed", location.getSpeed());

        positionProperties.putMap("coords", coords);
        positionProperties.putDouble("timestamp", location.getTime());
        return positionProperties;
    }
}

File: android/rctmgl/src/main/java-v10/com/mapbox/rctmgl/components/camera/RCTMGLCamera.kt
package com.mapbox.rctmgl.components.camera

import android.animation.Animator
import android.content.Context
import android.location.Location
import com.facebook.react.bridge.ReadableMap
import com.mapbox.maps.plugin.gestures.gestures
import com.mapbox.rctmgl.location.LocationManager.Companion.getInstance
import com.mapbox.maps.plugin.animation.flyTo
import com.mapbox.rctmgl.components.AbstractMapFeature
import com.mapbox.rctmgl.components.mapview.RCTMGLMapView
import com.mapbox.rctmgl.components.location.LocationComponentManager
import com.mapbox.rctmgl.utils.LatLngBounds
import com.mapbox.maps.plugin.locationcomponent.OnIndicatorBearingChangedListener
import com.mapbox.maps.plugin.locationcomponent.OnIndicatorPositionChangedListener
import com.facebook.react.bridge.WritableMap
import com.facebook.react.bridge.WritableNativeMap
import com.mapbox.maps.*
import com.mapbox.maps.plugin.PuckBearingSource
import com.mapbox.maps.plugin.locationcomponent.location
import com.mapbox.maps.plugin.locationcomponent.location2
import com.mapbox.maps.plugin.viewport.ViewportStatus
import com.mapbox.maps.plugin.viewport.ViewportStatusObserver
import com.mapbox.maps.plugin.viewport.data.FollowPuckViewportStateBearing
import com.mapbox.maps.plugin.viewport.data.FollowPuckViewportStateOptions
import com.mapbox.maps.plugin.viewport.data.ViewportStatusChangeReason
import com.mapbox.maps.plugin.viewport.state.FollowPuckViewportState
import com.mapbox.maps.plugin.viewport.state.OverviewViewportState
import com.mapbox.maps.plugin.viewport.state.ViewportState
import com.mapbox.maps.plugin.viewport.viewport
import com.mapbox.rctmgl.components.RemovalReason
import com.mapbox.rctmgl.components.camera.constants.CameraMode
import com.mapbox.rctmgl.components.location.*
import com.mapbox.rctmgl.events.MapUserTrackingModeEvent
import com.mapbox.rctmgl.location.*
import com.mapbox.rctmgl.utils.Logger
import com.mapbox.rctmgl.utils.writableMapOf


class RCTMGLCamera(private val mContext: Context, private val mManager: RCTMGLCameraManager) :
    AbstractMapFeature(
        mContext
    ) {
    private var hasSentFirstRegion = false
    private var mDefaultStop: CameraStop? = null
    private var mCameraStop: CameraStop? = null
    private val mCameraUpdateQueue = CameraUpdateQueue()

    /*
    // private LocationComponent mLocationComponent;
     */
    private var mLocationComponentManager: LocationComponentManager? = null
    private var mUserTrackingMode = 0
    private var mUserTrackingState = UserTrackingState.POSSIBLE
    private val mLocationManager: LocationManager?
    private val mUserLocation: UserLocation = UserLocation()
    private val mCenterCoordinate: ScreenCoordinate? = null
    private val mAnimated = false
    private val mHeading = 0.0

    private var mFollowUserLocation = false
    private var mFollowUserMode: String? = null
    private var mFollowZoomLevel : Double? = null
    private var mFollowPitch : Double? = null
    private var mFollowHeading : Double? = null
    private var mFollowPadding : EdgeInsets? = null

    private var mZoomLevel = -1.0
    private var mMinZoomLevel : Double? = null
    private var mMaxZoomLevel : Double? = null
    private var mMaxBounds: LatLngBounds? = null


    private val mCameraCallback: Animator.AnimatorListener = object : Animator.AnimatorListener {
        override fun onAnimationStart(animator: Animator) {}
        override fun onAnimationEnd(animator: Animator) {
            if (!hasSentFirstRegion) {
                mMapView?.sendRegionChangeEvent(false)
                hasSentFirstRegion = true
            }
        }

        override fun onAnimationCancel(animator: Animator) {
            if (!hasSentFirstRegion) {
                mMapView?.sendRegionChangeEvent(false)
                hasSentFirstRegion = true
            }
        }

        override fun onAnimationRepeat(animator: Animator) {}
    }

    override fun addToMap(mapView: RCTMGLMapView) {
        super.addToMap(mapView)
        setInitialCamera()
        updateMaxBounds()
        mCameraStop?.let { updateCamera(it) }
        _observeViewportState(mapView.mapView)
        _updateViewportState()
    }

    override fun removeFromMap(mapView: RCTMGLMapView, reason: RemovalReason) : Boolean {
        if (reason == RemovalReason.STYLE_CHANGE) {
            return false
        } else {
            return super.removeFromMap(mapView, reason);
        }
    }
    fun setStop(stop: CameraStop) {
        mCameraStop = stop
        stop.setCallback(mCameraCallback)
        if (mMapView != null) {
            stop.let { updateCamera(it) }
        }
    }

    fun setDefaultStop(stop: CameraStop?) {
        mDefaultStop = stop
    }

    fun setFollowUserMode(mode: String?) {
        mFollowUserMode = mode
        _updateViewportState()
    }

    fun setFollowUserLocation(value: Boolean) {
        mFollowUserLocation = value
        _updateViewportState()
    }

    fun setFollowZoomLevel(zoomLevel: Double) {
        mFollowZoomLevel = zoomLevel
        _updateViewportState();
    }

    fun setFollowPitch(pitch: Double) {
        mFollowPitch = pitch
        _updateViewportState();
    }

    fun setFollowHeading(heading: Double) {
        mFollowHeading = heading
        _updateViewportState();
    }

    fun setFollowPadding(padding: ReadableMap) {
        // scale padding by pixel ratio
        val metrics = context.resources.displayMetrics
        val edgeInsets = EdgeInsets(
            if (padding.hasKey("paddingTop")) padding.getDouble("paddingTop") * metrics.density else 0.0,
            if (padding.hasKey("paddingLeft")) padding.getDouble("paddingLeft") * metrics.density else 0.0,
            if (padding.hasKey("paddingBottom")) padding.getDouble("paddingBottom") * metrics.density else 0.0,
            if (padding.hasKey("paddingRight")) padding.getDouble("paddingRight") * metrics.density else 0.0,
        )

        mFollowPadding = edgeInsets
        _updateViewportState();
    }

    fun setMaxBounds(bounds: LatLngBounds?) {
        mMaxBounds = bounds
        updateMaxBounds()
    }


    private fun updateMaxBounds() {
        withMapView { mapView ->
            val map = mapView.getMapboxMap()
            val maxBounds = mMaxBounds
            val builder = CameraBoundsOptions.Builder()

            if (maxBounds != null) {
                builder.bounds(maxBounds.toBounds())
            }
            mMinZoomLevel?.let { builder.minZoom(it) }
            mMaxZoomLevel?.let { builder.maxZoom(it) }
            map.setBounds(builder.build())
        }
    }

    private fun setInitialCamera() {
        mDefaultStop?.let {
            val mapView = mMapView!!
            val map = mapView.getMapboxMap()

            it.setDuration(0)
            it.setMode(CameraMode.NONE)
            val item = it.toCameraUpdate(mapView)
            item.run()
        }
    }

    private fun updateCamera(cameraStop: CameraStop) {
        mCameraUpdateQueue.offer(cameraStop)
        mCameraUpdateQueue.execute(mMapView)
    }

    private fun updateUserLocation(isAnimated: Boolean) {

    }

    // NOTE: The direction of this is used for map rotation only, not location layer rotation
    private val directionForUserLocationUpdate: Double
        private get() {
            // NOTE: The direction of this is used for map rotation only, not location layer rotation
            val currentCamera = mMapView!!.getMapboxMap().cameraState
            var direction = currentCamera.bearing
            val userTrackingMode = mUserLocation.trackingMode
            if (userTrackingMode == UserTrackingMode.FollowWithHeading || userTrackingMode == UserTrackingMode.FollowWithCourse) {
                direction = mUserLocation.bearing
            } else if (mHeading != 0.0) {
                direction = mHeading
            }
            return direction
        }

    private fun hasSetCenterCoordinate(): Boolean {
        val state = mapboxMap!!.cameraState
        val center = state.center
        return center.latitude() != 0.0 && center.longitude() != 0.0
    }

    init {
        mLocationManager = getInstance(mContext)
    }

    private fun updateLocationLayer(style: Style) {
        if (mLocationComponentManager == null) {
            mLocationComponentManager = mMapView!!.locationComponentManager
        }
        mLocationComponentManager!!.update()
    }

    fun setMinZoomLevel(zoomLevel: Double?) {
        mMinZoomLevel = zoomLevel
        updateMaxBounds()
    }

    fun setMaxZoomLevel(zoomLevel: Double?) {
        mMaxZoomLevel = zoomLevel
        updateMaxBounds()
    }

    fun setZoomLevel(zoomLevel: Double) {
        mZoomLevel = zoomLevel
        updateCameraPositionIfNeeded(false)
    }

    private fun buildCamera(
        previousPosition: CameraState,
        shouldUpdateTarget: Boolean
    ): CameraOptions {
        return if (shouldUpdateTarget) {
            previousPosition.toCameraOptions(mCenterCoordinate)
        } else {
            previousPosition.toCameraOptions(null)
        }
    }

    private fun updateCameraPositionIfNeeded(shouldUpdateTarget: Boolean) {
        if (mMapView != null) {
            val prevPosition = mapboxMap!!.cameraState
            val cameraUpdate =  /*CameraUpdateFactory.newCameraPosition(*/
                buildCamera(prevPosition, shouldUpdateTarget)
            if (mAnimated) {
                mapboxMap!!.flyTo(cameraUpdate, null)
            } else {
                mapboxMap!!.setCamera(cameraUpdate)
            }
        }
    }

    fun setUserTrackingMode(userTrackingMode: Int) {
        val oldTrackingMode = mUserTrackingMode
        mUserTrackingMode = userTrackingMode
        mManager.handleEvent(MapUserTrackingModeEvent(this@RCTMGLCamera, userTrackingMode))
        when (mUserTrackingMode) {
            UserTrackingMode.NONE -> mUserTrackingState = UserTrackingState.POSSIBLE
            UserTrackingMode.FOLLOW, UserTrackingMode.FollowWithCourse, UserTrackingMode.FollowWithHeading -> if (oldTrackingMode == UserTrackingMode.NONE) {
                mUserTrackingState = UserTrackingState.POSSIBLE
            }
        }
        if (mapboxMap != null) {
            updateLocationLayer(mapboxMap!!.getStyle()!!)
        }
    }

    fun toFollowUserLocation(toStatus: ViewportStatus): Boolean {
        when (toStatus) {
            ViewportStatus.Idle -> return false
            is ViewportStatus.State -> return true
            is ViewportStatus.Transition -> return true
        }
    }

    fun toFollowUserMode(state: ViewportState): String? {
        if (state is FollowPuckViewportState) {
            when (state.options.bearing) {
                is FollowPuckViewportStateBearing.SyncWithLocationPuck ->
                    return "normal"
                is FollowPuckViewportStateBearing.Constant ->
                    return "constant"
                else -> {
                    Logger.w(LOG_TAG, "Unexpected bearing: ${state.options.bearing}")
                    return "normal"
                }
            }
        } else if (state is OverviewViewportState) {
            return "overview"
        } else {
            return "custom"
        }
    }

    fun toFollowUserMode(status: ViewportStatus): String? {
        when (status) {
            ViewportStatus.Idle -> return null
            is ViewportStatus.State ->
                return toFollowUserMode(status)
            is ViewportStatus.Transition ->
                return toFollowUserMode(status.toState)
        }
    }

    fun toReadableMap(status: ViewportStatus): ReadableMap {
        when (status) {
            ViewportStatus.Idle -> return writableMapOf("state" to "idle")
            is ViewportStatus.State ->
                return writableMapOf(
                    "state" to status.toString()
                )
            is ViewportStatus.Transition ->
                return writableMapOf(
                    "transition" to status.toString()
                )
        }
    }

    fun _observeViewportState(mapView: MapView) {
        mapView.viewport.addStatusObserver(object: ViewportStatusObserver {
            override fun onViewportStatusChanged(
                fromStatus: ViewportStatus,
                toStatus: ViewportStatus,
                reason: ViewportStatusChangeReason
            ) {
                if (reason == ViewportStatusChangeReason.USER_INTERACTION) {
                    val followUserLocation = toFollowUserLocation(toStatus)



                    mManager.handleEvent(MapUserTrackingModeEvent(this@RCTMGLCamera, UserTrackingMode.NONE,
                    writableMapOf(
                        "followUserMode" to toFollowUserMode(toStatus),
                        "followUserLocation" to followUserLocation,
                        "fromViewportState" to toReadableMap(fromStatus),
                        "toViewportState" to toReadableMap(toStatus),
                        "reason" to toString(reason),
                    )
                    ))
                }
            }
        })
    }

    fun toString(reason: ViewportStatusChangeReason): String {
        when (reason) {
            ViewportStatusChangeReason.IDLE_REQUESTED ->
                return "idleRequested"
            ViewportStatusChangeReason.TRANSITION_FAILED ->
                return "transitionFailed"
            ViewportStatusChangeReason.TRANSITION_STARTED ->
                return "transitionStarted"
            ViewportStatusChangeReason.TRANSITION_SUCCEEDED ->
                return "transitionSucceeded"
            ViewportStatusChangeReason.USER_INTERACTION ->
                return "userInteraction"
            else -> {
                Logger.w(LOG_TAG, "toString; unkown reason: ${reason}")
                return "unkown: $reason"
            }
        }
    }

    fun _updateViewportState() {
        mMapView?.let {
            val map = it.mapView
            val viewport = map.viewport;

            if (mLocationComponentManager == null) {
                mLocationComponentManager = it.locationComponentManager
            }

            if (mFollowUserLocation == false) {
                viewport.idle()
                mLocationComponentManager?.setFollowLocation(false)
                return;
            }

            mLocationComponentManager?.setFollowLocation(true)
            mLocationManager?.let {
                val provider = map.location.getLocationProvider()
                if (provider != null) {
                    it.provider = provider
                }
            }

            val location = map.location2
            val followOptions = FollowPuckViewportStateOptions.Builder()
            val cameraState = map.getMapboxMap().cameraState
            when (mFollowUserMode ?: "normal") {
                "compass" -> {
                    location.puckBearingEnabled = true
                    location.puckBearingSource = PuckBearingSource.HEADING
                    followOptions.bearing(FollowPuckViewportStateBearing.SyncWithLocationPuck)
                }
                "course" -> {
                    location.puckBearingEnabled = true
                    location.puckBearingSource = PuckBearingSource.COURSE
                    followOptions.bearing(FollowPuckViewportStateBearing.SyncWithLocationPuck)
                }
                "normal" -> {
                    location.puckBearingEnabled = false
                    when(val it=mFollowHeading) {
                        null -> followOptions.bearing( FollowPuckViewportStateBearing.Constant(
                            cameraState.bearing
                        ))
                        else -> followOptions.bearing( FollowPuckViewportStateBearing.Constant(
                            cameraState.bearing
                        ))
                    }
                }
                else -> {
                    Logger.e("RCTMGLCamera", "unexpected follow mode: $mFollowUserMode")
                }
            }

            when(val it=mFollowZoomLevel) {
                null -> followOptions.zoom(cameraState.zoom)
                else -> followOptions.zoom(it)
            }

            when(val it=mFollowPitch) {
                null -> followOptions.pitch(cameraState.pitch)
                else -> followOptions.pitch(it)
            }

            when(val it=mFollowPadding) {
                null -> followOptions.padding(cameraState.padding)
                else -> followOptions.padding(it)
            }


            val followState = viewport.makeFollowPuckViewportState(followOptions.build())
            viewport.transitionTo(followState)
        }
        mapboxMap?.let {
            it.getStyle()?.let {
                updateLocationLayer(it)
            }
        }
    }

    private fun updatedFollowUserMode() {
        if (mFollowUserLocation) {
            setUserTrackingMode(UserTrackingMode.fromString(mFollowUserMode))
        } else {
            setUserTrackingMode(UserTrackingMode.NONE)
        }
    }

    val mapboxMap: MapboxMap?
        get() = if (mMapView == null) {
            null
        } else mMapView!!.getMapboxMap()

    /**
     * Create a payload of the location data per the web api geolocation spec
     * https://dev.w3.org/geo/api/spec-source.html#position
     *
     * @return
     */
    private fun makeLocationChangePayload(location: Location): WritableMap {
        val positionProperties: WritableMap = WritableNativeMap()
        val coords: WritableMap = WritableNativeMap()
        coords.putDouble("longitude", location.longitude)
        coords.putDouble("latitude", location.latitude)
        coords.putDouble("altitude", location.altitude)
        coords.putDouble("accuracy", location.accuracy.toDouble())
        // A better solution will be to pull the heading from the compass engine, 
        // unfortunately the api is not publicly available in the mapbox sdk
        coords.putDouble("heading", location.bearing.toDouble())
        coords.putDouble("course", location.bearing.toDouble())
        coords.putDouble("speed", location.speed.toDouble())
        positionProperties.putMap("coords", coords)
        positionProperties.putDouble("timestamp", location.time.toDouble())
        return positionProperties
    }

    companion object {
        const val USER_LOCATION_CAMERA_MOVE_DURATION = 1000
        const val minimumZoomLevelForUserTracking = 10.5
        const val defaultZoomLevelForUserTracking = 14.0
        const val LOG_TAG = "RCTMGLCamera"
    }
}